### PR TITLE
JDK-8252717: Integrate/merge legacy standard doclet diagnostics and doclint

### PR DESCRIPTION
**Pull request by jonathan-gibbons**
Please review a moderate change to enable (most) doclet warnings even if doclint is enabled.

Since the introduction of doclint, there was some (small) overlap between the small set of warnings generated by the doclet and the new larger set of diagnostics that could be generated by doclint.  The solution, until now, has been to disable doclet warnings when doclint is enabled. But, the sets do not overlap, and the policy has inappropriately suppressed some warnings and inhibited the generation of new warnings by the doclet that could only be done by the doclet, and not doclint.

One notable group of warnings that has been inappropriately suppressed is the warnings generated by using the `-serial warn` option.

The fundamental core of the change is to remove the conditional checks in the doclet `Messages.java`, which would suppress messages when doclint was enabled.  A consequence is that some specific messages have to disabled if they are duplicate checks if doclint is enabled. (The messages cannot be removed altogether because doclint might _not_ be enabled.)   A test is added to verify that either one message or the other is generated, but never both.

As previously noted, an issue with the earlier policy is that warnings generated by using the `-serial warn` option were inappropriately suppressed, and this change fixes that ... revealing a number of latent warnings in the JDK API that need to be addressed.  The short term fix here is to temporarily remove the use of the `-serialwarn` option. See [JDK-8287749](https://bugs.openjdk.java.net/browse/JDK-8287749).

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.java.net/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8252717](https://bugs.openjdk.java.net/browse/JDK-8252717): Integrate/merge legacy standard doclet diagnostics and doclint

### Reviewers
 * [Erik Joelsson](https://openjdk.java.net/census#erikj) (@erikj79 - **Reviewer**) ⚠️ Review applies to [2f4bc7fb](https://git.openjdk.java.net/jdk/pull/9006/files/2f4bc7fb28be7ed55f542049658e4bba02224775)
 * [Pavel Rappo](https://openjdk.java.net/census#prappo) (@pavelrappo - **Reviewer**) ⚠️ Review applies to [2f4bc7fb](https://git.openjdk.java.net/jdk/pull/9006/files/2f4bc7fb28be7ed55f542049658e4bba02224775)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/9006/head:pull/9006` \
`$ git checkout pull/9006`

Update a local copy of the PR: \
`$ git checkout pull/9006` \
`$ git pull https://git.openjdk.java.net/jdk pull/9006/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 9006`

View PR using the GUI difftool: \
`$ git pr show -t 9006`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/9006.diff">https://git.openjdk.java.net/jdk/pull/9006.diff</a>

</details>

### CODE REVIEW

**Review by erikj79 — APPROVED** [review]
Build change looks ok.

**Review by pavelrappo — COMMENTED** [review]
> Since the introduction of doclint, there was some (small) overlap between the small set ... and the new larger set ... But, the sets do not overlap

Do or do they not overlap? :)

> the policy has inappropriately suppressed some warnings and inhibited the generation of new warnings

What's the difference between "suppressed" and "inhibited" here?

Generally, how can we be sure that we haven't missed any double-reported cases?

**Comment by pavelrappo** [comment]
> I did a visual inspection over the Checker class.

> The test is set up to be a framework to accommodate new cases if any should arise.

This is good. It would be even better if we had a mechanism to reduce risk of diagnostic duplication in the future. I don't have any concrete proposal, but unique internal error codes shared between doclint and doclet come to mind.

**Review by pavelrappo — APPROVED** [review]
Looks good.

**Comment by jonathan-gibbons** [comment]
> > I did a visual inspection over the Checker class.
> 
> > The test is set up to be a framework to accommodate new cases if any should arise.
> 
> This is good. It would be even better if we had a mechanism to reduce risk of diagnostic duplication in the future. I don't have any concrete proposal, but unique internal error codes shared between doclint and doclet come to mind.

I would also like to have a policy to align the user-facing content of the diagnostic message. This includes the severity (warning or error) as well as the position and message text, including substituted values.

File: test/langtools/jdk/javadoc/doclet/testDoclintDocletMessages/TestDocLintDocletMessages.java [hunk 1]
```diff
@@ -0,0 +1,197 @@
...
+        var allOptions = new ArrayList<String>();
+        allOptions.addAll(List.of("-d", base.resolve("out-" + index).toString()));
+        allOptions.addAll(options);
+        allOptions.addAll(List.of("-noindex", "-nohelp")); // omit unnecessary files
+        allOptions.add(base.resolve("src").resolve("C.java").toString());
+
+        javadoc(allOptions.toArray(String[]::new));
+        checkExit(expect.exit);
+
+        checkOutput(Output.OUT, true,expect.message);
```
[the remark this answers is not in the document]
Comment (jonathan-gibbons) [reply test/langtools/jdk/javadoc/doclet/testDoclintDocletMessages/TestDocLintDocletMessages.java]: ✅

**Pull request closed without merging**